hydrogym/ts.py

```diff
@@ -284,6 +285,58 @@ def rmatvec(q_vec):
                 B[:, i] = get_array(Bi)
             return A, B
 
+class Torque_IPCS(IPCS):
+    def __init__(self, flow: FlowConfig, dt: float, debug=False, **kwargs):
+        super().__init__(flow, dt, debug=False, **kwargs)
+        self.control = 0.0 # init with no torque on system
+        self.omega = [0.0] # Cylinder starts at rest
+        self.k_damping = 1/self.flow.TAU
+        self.I_cm = 0.006172 # Moment of inertia about CoM of a Plexiglass Cylinder with a 2 inch radius and spanning a half meter test section of a wind tunnel
+
+    def set_dampingConst(self, k_new):
+        self.k_damping = k_new
+        return
+
+    def get_momOfInertia(self):
+        return self.I_cm
+
+    def get_omega(self):
+        return self.omega
+
+    def back_euler_func(self, x):
+        return x + ((self.k_damping * x - self.control)/self.I_cm) * self.dt - self.omega
+
+    def update_velocity(self, torque):
+        """
+        Update BCS of cylinder to new angular velocity
+        """
+        self.omega = fsolve(self.back_euler_func, self.omega)
+        return
+
+    def step(self, iter, control=None):
```

Comment:
The difference is in the implementation of `update_controls/update_velocity`.

`update_controls` assumes the control is the same as the state which is incompatible with the way `update_velocity` updates the state based on the control. If we were to merge IPCS_torque into IPCS, I believe this would only be possible via an indicator of whether torque or velocity is the control variable. Something like:
`
class IPCS(TransientSolver):
    def __init__(self, ..., control_variable='velocity', ...):
            super().__init__(flow, dt)
            ...
            self.control_variable = control_variable # either 'velocity' or 'torque'
`
Where later in the code we merge `update_controls/update_velocity` like so:

`def update_controls(self, control):
        if self.control_variable == 'torque':
            ...
            self.state = (self.state + self.control * dt / self.I_cm) / (1 + self.k_damping * dt / self.Icm)
        else:
            ...
            for (u, v) in zip(self.state, state):
                u = u + (self.dt/self.flow.TAU)*(v - u)

`

Then we would always just have `self.state` = `self.control` for the case where velocity is the control variable, which is a bit redundant but easy to do (or we could just ignore that self.state exists when using velocity as the control).

I do see the advantage to merging the two into one class though, so I will try and get something together over the next couple days while changing as little of the original code as possible.

Reply:
@jcallaham let me know what you think of the above proposal to have an indicator on "what the control variable is". Also to make this more general, I should use other keywords besides 'torque' and 'velocity', but the general framework should be the same. 

Essentially, I would have a "directly controlling state_IPCS" where the control allows you to directly input the state that you want and an "input based_IPCS" similar to Torque_IPCS where the control is something that affects the current state of the environment indirectly. Then the "directly controlling state_IPCS" would be activated with one keyword and the "input based _IPCS" would be activated with a different keyword. I think this should work, but we'll see!

Reply:
Yeah that's exactly the structure I had in mind.  Eventually we might even get rid of the "direct" version and just make sure that the default behavior is somehow equivalent (and that the "indirect" version is differentiable", but I like the backwards compatibility.  I actually liked "velocity" and "force" as keywords better than "direct" and "indirect" but otherwise I think this is a nice way of incorporating the functionality into the existing classes.